### PR TITLE
@CompileStatic with exclusions.

### PR DESCRIPTION
**Pull request by javabrett**
Re #46.  Not for merging yet but posting branch and PR to aid discussion.

### CODE REVIEW

**Comment by javabrett** [comment]
@boris-petrov yes they were added one-by-one until the project compiled.

File: gradle/config.groovy [hunk 1]
```diff
@@ -0,0 +1,3 @@
+withConfig(configuration) {
```
Comment (boris-petrov) [inline gradle/config.groovy]: I'm not exactly sure what this does, but [here](https://stackoverflow.com/a/38620029/4478636) is another way of doing the same as what you've done in `common.gradle` and `gradle/config.groovy`. I think?

Comment (javabrett) [reply gradle/config.groovy]: Right, that link shows both methods of adding the compiler configuration script elements required.  I believe that the option to put them in a single file is newer.

For some reason a single file containing all the compiler-directives appealed to me, and I found the AST version easier to read.  Do you have a strong-preference for the scripted version, eliminating `config.groovy`?

Comment (boris-petrov) [reply gradle/config.groovy]: No strong preference at all. :) Just less code/files but you are right about having all configurations in a single place. So leave it like that.

Comment (javabrett) [reply gradle/config.groovy]: It's probably moot anyway based on the discussion over in the original issue ... making everything `@CompileStatic` is being discussed as overkill.  Apparently the only thing that will make any difference is Gradle configuration-time activity, when the plugin is applied, not when it executes.

Comment (boris-petrov) [reply gradle/config.groovy]: Well, in our project we have a lot of Groovy and at some point I decided I hated the performance and went ahead and added `CompileStatic` *everywhere* and it ran **much** faster. I don't see any reason for not adding it absolutely everywhere (apart where heavy meta-programming or reflection is used) even if that would make only the compile-time faster (which is a win again).

Comment (javabrett) [reply gradle/config.groovy]: > and went ahead and added CompileStatic everywhere

Interesting - does that mean that your project was pretty-much Java-code but compiled with Groovy?  Based on my first experience here trying to add `@CompileStatic` in Gretty - pretty much any syntax which is not strict Java or anything with any dynamic typing at all immediately fails the static-compile, unsurprisingly.  Each of the files I've had to exclude should have at least one example of something that can only be done with Groovy dynamic allowances.

I'm guessing it's not going to be helpful or fun converting a bunch of those to work in static Java compilation, even for the ones that are "reasonable" changes.

Comment (boris-petrov) [reply gradle/config.groovy]: We have mostly Java code with Groovy for tests and some other minor stuff - so it wasn't too much non-Java code. Yes, we had to change the Groovy (perhaps even a lot in some places - but mostly adding type annotations) but in the end it was worth it. I'm not sure here it will be - probably it will be too much work, not sure.

I guess what you have done here suffices for now. Although I'm not exactly sure how the Groovy compiler can generate "fast" bytecode when it doesn't know the types... :)

**Comment by boris-petrov** [comment]
@javabrett - why did you close this?

**Comment by javabrett** [comment]
> @javabrett - why did you close this?

I'd prefer to make this change (adding some `@CompileStatic`) mostly if it can be shown to be beneficial - some tangible benefit.

Originally raised in https://github.com/akhikhl/gretty/issues/428 the assertion is that [Spring Security](https://github.com/spring-projects/spring-security) ran gradle-profiler ([results](https://scans.gradle.com/s/7yjns4i55vj66/)), and there's a claim that:

> The vast majority of the time (75%) is spent in the Gretty plugin, which is doing a lot of heavy lifting in dynamic Groovy. If it was migrated to @compilestatic (or Java or Kotlin) you’d see a big improvement in your configuration time.

I have since run the gradle-profiler for the same Spring Security project, but I'm not yet able to interpret the results to draw the same conclusion.  So I've asked for some help on that issue.

In short: I'd like to make this change with a measurable test-case to match.  This first broad-brush approach was just to see if some massive performance difference emerged and it did not.  Subsequent advice was that we don't really need to add the annotation that broadly to fix the problem.

I'd like to focus on a reproducible test first.

**Comment by javabrett** [comment]
Reopening and changing base-branch to `gretty-2.x`.

**Comment by javabrett** [comment]
Fixed on branch `gretty-2.x` with 3cb25470694d86898f8e72df9da2b05c6766d3db .
